### PR TITLE
Connecting block to parent in domToBlockHeadless_

### LINKED ISSUE #2135: Dynamic dropdown "forgets" option keys when reloaded
Given a dynamic dropdown, i.e. one where `options` is set to a function instead of a fixed array, if that function returns a list of options where the option key is different to the option value, the option key is lost upon reloading the workspace.

### Expected Behavior

A dynamic dropdown should always show the option key to the user after a workspace using it is reloaded, not the option value.

### Actual Behavior

Given a dynamic dropdown, if the workspace containing it is persisted to XML and then later reloaded, the option keys are lost and the option value is shown to the user in the workspace.

### Steps to Reproduce

Given a dynamic dropdown function like:

```
function dynamicOptions() {
   return [["a", "b"]];
```

If you have a block on your workspace that uses it, it will show `a` as the option to the user upon being dragged onto the workspace.

If you then write the workspace to XML with `Blockly.Xml`, create a new empty workspace and populate it from the XML (again with `Blockly.Xml`) it will show the same block with the same dropdown, but this time the option will be shown as `b` to the user, i.e. Blockly is treating the option value as the key for display purposes.

### Additional Information

From what I can see the issue is that the option keys are not represented in the workspace XML and the dropdown does not infer them from the dynamic option function when deserialized.

Strangely, if you open the dropdown to see the options, it highlights the correct one by key.

In my real life use case, the option values are a JSON datastructure, so this looks pretty ugly. Is there any way to force the dropdown to re-initialize the options upon deserialization?

### PR DESCRIPTION
**Pull request by moniika**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

Fixes #2135

### Proposed Changes

- Connects block to it's parent in domToBlockHeadless_ rather than in the parent block's instantiation (allows for parent to be connected before field tag is processed)
  - Before: Block was connected it it's parent after it was fully loaded (which caused issues with fields validators that depended on the connected state of the block)
  - After: Block is connected to parent as it is being loaded, before it's field values are loaded

### Reason for Changes

- Solves issues with loading blocks with field validators that depend on the parent being connected (and accessing parent fields) from xml
- Solves issues with loading dynamic dropdowns that depend on a connected parent from xml

### Test Coverage

Ran tests and tested in advanced playground with the following block:
```
<xml xmlns="https://developers.google.com/blockly/xml">
  <variables>
    <variable>list</variable>
  </variables>
  <block type="controls_repeat_ext">
    <value name="TIMES">
      <block type="math_number">
        <field name="NUM">123</field>
      </block>
    </value>
    <statement name="DO">
      <block type="test_dropdowns_dynamic_connect_dependant">
        <field name="OPTIONS">wcjS31{oxnJqsPUgQ+_m_type_key</field>
        <next>
          <block type="lists_getIndex">
            <mutation statement="true" at="true"></mutation>
            <field name="MODE">REMOVE</field>
            <field name="WHERE">FROM_START</field>
            <value name="VALUE">
              <block type="variables_get">
                <field name="VAR">list</field>
              </block>
            </value>
            <value name="AT">
              <block type="lists_getIndex">
                <mutation statement="false" at="true"></mutation>
                <field name="MODE">GET</field>
                <field name="WHERE">FROM_START</field>
                <value name="VALUE">
                  <block type="variables_get">
                    <field name="VAR">list</field>
                  </block>
                </value>
                <value name="AT">
                  <block type="math_number">
                    <field name="NUM">123</field>
                  </block>
                </value>
              </block>
            </value>
            <next>
              <block type="test_dropdowns_dynamic_connect_dependant">
                <field name="OPTIONS">T1_y8l67gg$DXQ=u8Q~2_type_key</field>
              </block>
            </next>
          </block>
        </next>
      </block>
    </statement>
  </block>
</xml>
```
### Additional Information

The generator tests were very helpful for testing this change as it caught many edge cases of xml generation.

Depends on https://github.com/google/blockly/pull/4571

### CODE REVIEW

**Comment by BeksOmega** [comment]
With the new connection checker being more flexible, I wonder about connection checks that depend on the state of a parent block's fields. Wouldn't this mess with that?

File: core/xml.js [hunk 1]
```diff
@@ -631,125 +634,164 @@ Blockly.Xml.domToBlockHeadless_ = function(xmlBlock, workspace) {
...
-        }
+        xmlChildNameMap.next.push(xmlChild);
         break;
       default:
         // Unknown tag; ignore.  Same principle as HTML parsers.
         console.warn('Ignoring unknown tag: ' + xmlChild.nodeName);
     }
   }
 
+  var callInitSvg = false;
```
Comment (rachel-fenichel) [inline core/xml.js:687]: Break this out into helper functions where possible.

File: core/xml.js [hunk 2]
```diff
@@ -631,125 +634,164 @@ Blockly.Xml.domToBlockHeadless_ = function(xmlBlock, workspace) {
...
+   * A mapping of nodeName to node for child nodes of xmlBlock.
+   * @type {{next: Array<Element>, mutation: Array<Element>,
+   *    input: Array<Element>, data: Array<Element>, field: Array<Element>,
+   *    comment: Array<Element>}}
+   */
+  var xmlChildNameMap = {
+    mutation: [], comment: [], data: [], field: [], input: [],
+    next: []
+  };
   for (var i = 0, xmlChild; (xmlChild = xmlBlock.childNodes[i]); i++) {
```
Comment (rachel-fenichel) [inline core/xml.js:648]: This loop can all be pulled out into a helper function that returns the generated map.

File: package.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
     "recompile": "gulp gitPreCompile && npm run bump && gulp gitPostCompile",
     "release": "gulp gitCreateRC",
     "test": "concurrently 'npm run test:prepare' 'sleep 5 && npm run test:run'",
+    "test:generators": "concurrently 'npm run test:prepare' 'sleep 25 && tests/scripts/run_generators.sh'",
```
Comment (rachel-fenichel) [inline package.json:40]: Why is the sleep now 25?

**Review by rachel-fenichel — COMMENTED** [review]
Some comments.

I'm nervous about this because of backwards compatibility, but agree that making our parsing more clearly defined is a good move.

File: core/xml.js (hunk 2 above)
Comment (moniika) [reply core/xml.js:648]: Done.

File: core/xml.js [hunk 4]
```diff
@@ -631,125 +634,164 @@ Blockly.Xml.domToBlockHeadless_ = function(xmlBlock, workspace) {
   var id = xmlBlock.getAttribute('id');
   block = workspace.newBlock(prototypeName, id);
 
-  var blockChild = null;
+  // Preprocess childNodes so tags can be processed in a consistent order.
+  /**
+   * A mapping of nodeName to node for child nodes of xmlBlock.
+   * @type {{next: Array<Element>, mutation: Array<Element>,
```
[the remark this answers is not in the document]
Comment (moniika) [reply core/xml.js:640]: Done

File: core/xml.js (hunk 1 above)
Comment (moniika) [reply core/xml.js:687]: Done

File: package.json (hunk 3 above)
Comment (moniika) [reply package.json:40]: At some point while i was testing the extra sleep was necessary for running generator tests. However, it looks like 5 seconds works for me locally, so I will change to 5.

File: core/xml.js [hunk 5]
```diff
@@ -619,10 +619,13 @@ Blockly.Xml.domToVariables = function(xmlVariables, workspace) {
  * workspace.
  * @param {!Element} xmlBlock XML block element.
  * @param {!Blockly.Workspace} workspace The workspace.
+ * @param {?Blockly.Connection=} parentConnection The parent connection to
```
[the remark this answers is not in the document]
Comment (moniika) [reply core/xml.js]: Updated.

**Comment by moniika** [comment]
Rebased after merging change for parsing tags in explicit order.

**Comment by moniika** [comment]
> With the new connection checker being more flexible, I wonder about connection checks that depend on the state of a parent block's fields. Wouldn't this mess with that?

I don't think this would be an issue if the connection checker depends on the **parent** block's fields.
The xml (#4571) is loaded in an explicit order, such that, the input/statement/value fields that create the child block are processed after the field tags are processed.

However, if the connection depends on the **child** block's fields, then those wont be loaded yet. 
(Ex: a block with boolean field that must be true in order for it's output to be connected to parent)
(Ex: a block whose field value must match the field value of the parent in order for it to be connected <- in previous release, this case only worked if the field in question was listed before the child block in xml)

Other potential **new** issue (introduced in #4571):
- field validator depends on connected child blocks (used to work **only if** the field tag was listed before the statement/value/input tag in the xml)

Other Remaining unsolved issues (not new issues):
- child field validator depends on parent's other connected statement/input/value child blocks
- field validator depends on all blocks in the workspace being loaded

**Review by rachel-fenichel — APPROVED** [review]
